hw/ip/otbn/util/shared/information_flow_analysis.py

```diff
@@ -0,0 +1,598 @@
+        insn.mnemonic))
+
+
+def _get_iflow_cache_fetch(pc: int, constants: Dict[str, int],
+                           cache: IFlowCache) -> Optional[IFlowResult]:
+    '''Returns a matching result from the cache if there is one.
+
+    The cache format maps the PC to a list of cache entries, with each entry
+    holding the cached result as well as the values of constants that it
+    depends on. The entry is a match if and only if those constant values match
+    the current constants.
+    '''
+    if pc in cache:
+        for constant_deps, result in cache[pc]:
+            if all(
+                    constants.get(name, None) == value
+                    for name, value in constant_deps.items()):
+                return result
+    return None
+
+
+def _get_iflow_cache_update(pc: int, constants: Dict[str, int],
+                            rec: IFlowResult, cache: IFlowCache) -> None:
+    '''Updates the cache for _get_iflow after a recursive call.'''
+    rec_used_constants = rec[0]
+    used_constant_values = {}
+    for name in rec_used_constants:
+        if name not in constants:
+            # Can't determine value; done
+            return
+        used_constant_values[name] = constants[name]
+
+    # Don't bother if an entry already exists
+    if _get_iflow_cache_fetch(pc, used_constant_values, cache) is not None:
+        return
+
+    if pc not in cache:
+        cache[pc] = []
+    cache[pc].append((used_constant_values, rec))
+    return
+
+
+def _get_iflow_update_state(
+        rec_result: IFlowResult, iflow: InformationFlowGraph,
+        program_end_iflow: Optional[InformationFlowGraph],
+        used_constants: Set[str],
+        control_deps: Dict[str, Set[int]]) -> Optional[InformationFlowGraph]:
+    '''Update the internal state of _get_iflow after a recursive call.
+
+    The `used_constants` and `control_deps` state elements are updated in
+    place, but the new `program_end_iflow` is returned. The `iflow` input is
+    not modified.
+    '''
+    rec_used_constants, _, rec_program_end_iflow, _, rec_control_deps = rec_result
+
+    # Update the used constants and control-flow dependencies
+    for const in rec_used_constants:
+        used_constants.update(iflow.sources(const))
+    for node, pcs in rec_control_deps.items():
+        # The dependencies for the caller are different from the ones in the
+        # recursive call; they are the *sources* of the recursive call's
+        # dependencies in the current information-flow graph.
+        for source_node in iflow.sources(node):
+            if source_node not in control_deps:
+                control_deps[source_node] = set()
+            control_deps[source_node].update(pcs)
+
+    # Update information flow results for paths where the program ends
+    if rec_program_end_iflow is not None:
+        end_iflow = iflow.seq(rec_program_end_iflow)
+        program_end_iflow = safe_update_iflow(program_end_iflow, end_iflow)
```

Comment:
This is actually a bit subtle. An empty graph says "nothing was updated in this path", which is different from "there's no path". An empty graph is equivalent to a graph with all nodes mapped to only themselves (i.e. their value at the end depends only on their value at the start).  Therefore, "empty SEQ G = G", but (more dangerously) "empty UPDATE G" is neither empty nor G.

For instance, if G has overwritten the value in `x3` so that it's a constant and no longer secret, then calling `update` will mean we're looking to combine all possible sources of information in `x3` in either G or the "empty" graph. But the empty graph says `x3`'s value stays the same as it was at the start;  in the empty graph, `x3` is still secret! So in the `update`d graph, `x3` will be marked as depending on its own initial value, and could be incorrectly listed as secret.

Reply:
Ah! Yep, that makes sense. Thanks for explaining!